### DISCUSSION TITLE
Async Operations in Python

### DISCUSSION THREAD

**Discussion by Isak-Bego** [AUBGTheHUB/spa-website-2022#774] [General]
Async/Await How does it work in python?
How does python deal with async operations?
What is the Event Loop?
How is the Event Loop blocked?

**Comment by Isak-Bego** [comment]
## Await/Async in Python

When we use Async in python we are creating a coroutine that can only be ran using await keyword. So if you have: 
```Python
from asyncio import run, sleep

async def hello1(): 
  sleep(2)
  print(Hello1!)

async def hello2(): 
  sleep(2)
  print(Hello2!)

async def main(): 
  hello1()
  hello2()


# You need to trigger the asyncio api in order to recognize the async and await keywords and to handle them accordingly
run(main())
```

This won't print anything because by calling the function like { name_function() },  we will receive a coroutine back, so we should in some way learn to deal with it. We can run the coroutine, which will trigger the function by awaiting it using the await keyword: 

```Python
from asyncio import run, sleep

async def hello1(): 
  sleep(2)
  print(Hello1!)

async def hello2(): 
  sleep(2)
  print(Hello2!)

async def main(): 
  await hello1()
  await hello2()


# You need to trigger the asyncio api in order to recognize the async and await keywords and to handle them accordingly
run(main())
```

```
# Console
2 seconds pass...
>>>Hello1!
2 seconds pass...
>>>Hello2!
```

COROUTINES are the essence of asyncronous code in python. [Learn More](https://www.geeksforgeeks.org/coroutine-in-python/) about them. 

However we did not get any concurrency action from this operation the coroutines ran squentially and behaved like syncronous code would behave. To enable these coroutines to run concurrently, we create what is called a task by passing the asyncio.create_task(coroutine=x). [Learn More](https://docs.python.org/3/library/asyncio-task.html) about tasks.

```Python
from asyncio import run, sleep, create_task

async def hello1(): 
  sleep(2)
  print(Hello1!)

async def hello2(): 
  sleep(2)
  print(Hello2!)

async def main(): 
  await create_task(hello1())
  await create_task(hello2())


# You need to trigger the asyncio api in order to recognize the async and await keywords and to handle them accordingly
run(main())
```

```
# Console
2 seconds pass...
>>>Hello1!
>>>Hello2!

# The events were ran concurrently
```

**Comment by Isak-Bego** [comment]
This is an amazing article to have a good grasp of how async works in python: 
https://medium.com/python-features/understanding-coroutines-tasks-in-depth-in-python-af2a4c0e1073
This is one of my favorite videos that explains the idea of how async is achieved in Javascript (top 3 fav vid in prog): 
https://www.youtube.com/watch?v=8aGhZQkoFbQ

**Comment by IvanObreshkov** [comment]
Here are my thoughts on the topic, though they might be a bit detailed:

---

### 1. Program Execution: From `python my_app.py` to the Event Loop

When you execute the command `python my_app.py` in a UNIX-like terminal (e.g., Linux, macOS), a precise sequence of events is initiated:

1.  **Process Creation:** The OS creates a new **Process**. A process is an isolated instance of a running program, given its own private virtual memory space by the kernel. This space contains the program's machine code, data, and a heap for dynamic objects.

2.  **Main Thread Spawning:** Within this new process, the OS starts a single **OS Thread**, known as the "main thread." A thread is the fundamental unit of execution that the OS scheduler dispatches to a CPU core. All execution for our program begins on this thread.

3.  **Code Loading & Interpretation:** The CPython interpreter (the running process) loads the contents of `my_app.py`. It compiles your Python source code into bytecode, which it prepares to execute on the main thread.

4.  **Event Loop Initialization:** Your code will contain an entry point like `asyncio.run(main())`.
    *   The `main()` call returns a **coroutine object**.
    *   The `asyncio.run()` function receives this object, instantiates an **Event Loop** (e.g., `SelectorEventLoop`), and binds it to the main thread.

5.  **The "Single-Threaded" Model:** `asyncio` enforces one running event loop per thread. Since a standard Python application has one main thread, a single event loop orchestrates all concurrent tasks. All your `async` Python code will execute on this one main thread, just at different, scheduled moments.

---

### 2. Definition of Python `asyncio` Terms and Data Structures

#### Terminology (The "Actors")

*   **Coroutine:** A special function defined with `async def`. Calling it returns a **coroutine object**, which is a pausable "recipe" for the function's execution. It only performs work when driven by an event loop, typically by being wrapped in a `Task`.
*   **Callback:** A regular Python function passed to the event loop to be executed later. It's the lowest-level scheduling mechanism.
*   **Handle:** An `asyncio.Handle` object that wraps a **Callback**. It's the "to-do item" placed in the `_ready` queue.
*   **TimerHandle:** A `Handle` with a `_when` attribute for timed execution.
*   **Task:** An `asyncio.Task` object that manages a **coroutine**. You create one with `asyncio.create_task(coro)`. It is a `Future` that runs its coroutine step-by-step.
*   **Future:** An `asyncio.Future`, a placeholder for a result that will be available later. It supports completion callbacks via `add_done_callback()`.

#### Data Structures (The "Locations")

*   **Call Stack:** Standard Python call stack on the main thread, tracking active function calls.
*   **Ready Queue (`loop._ready`):** A `collections.deque` of `Handle` objects ready to run *now*.
*   **Scheduled Queue (`loop._scheduled`):** A `heapq` of `TimerHandle` objects, ordered by wakeup time.
*   **Selector (`loop._selector`):** An object from the `selectors` module (e.g., `selectors.EpollSelector`) that wraps the OS's I/O multiplexing API (`epoll`, `kqueue`). It maintains the **I/O Waiter Map**, an internal mapping between a file descriptor and the **Callback** to run when that FD has an I/O event.

---

### 3. Types of Operations in the Event Loop

1.  **Timed Operations:** Managed by the **Scheduled Queue**.
2.  **Network I/O:** Managed by the **Selector**.
3.  **File I/O & Blocking Calls:** Managed by a **Worker Thread Pool**.

---

### 4. Detailed Execution Flow: Timed Operations

**Code Context:**
```python
import asyncio

async def my_main() -> None:
    print("Task: Starting...")
    # This is the line we are tracing
    await asyncio.sleep(0.1)
    print("Task: Finished.")

# This is where our trace begins
asyncio.run(my_main())
```

| Step | Action | `loop._ready` | `loop._scheduled` | `Task('my_main')` State/Properties | Code Context | Syscalls Used |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **1** | `asyncio.run(main())` creates a loop and calls `loop.create_task(my_main())`. | `deque([Handle(Task('my_main')._step)])` | `[]` | `PENDING` (`_state='PENDING'`) | `asyncio.run(my_main())` | - |
| **2** | `run()` starts the loop's `_run_once` cycle. It pops and executes `Handle(Task('my_main')._step)`. | `deque([])` | `[]` | `RUNNING` (`_state='RUNNING'`) | `my_main` starts, `print(...)` runs | - |
| **3** | `my_main` executes `await asyncio.sleep(0.1)`. This calls `loop.call_later()` which creates a `TimerHandle`. | `deque([])` | `[TimerHandle(when=100.1, ...)]` | `RUNNING` | `asyncio.sleep(0.1)` | `clock_gettime()` (or similar) |
| **4** | `await` pauses the `Task('my_main')` coroutine and **yields** control to the event loop. | `deque([])` | `[TimerHandle(when=100.1, ...)]` | `PENDING` (`_state='PENDING'`) | Control returns to `loop._run_once`. | - |
| **5** | `_run_once` finds `_ready` empty. It calls `loop._selector.select(timeout=0.1)`. The **main thread blocks**. | `deque([])` | `[TimerHandle(when=100.1, ...)]` | `PENDING` | Inside `loop._run_once` | `epoll_wait(timeout_ms=100)` |
| **6** | After 0.1s, the `epoll_wait` syscall times out. The main thread wakes up. | `deque([])` | `[TimerHandle(when=100.1, ...)]` | `PENDING` | Back in `loop._run_once`. | - |
| **7** | `_run_once` processes timers. The `TimerHandle` is now due. It's popped and its callback is wrapped in a `Handle` and added to `_ready`. | `deque([Handle(future.set_result)])` | `[]` | `PENDING` | `_run_once()` processes timers. | `clock_gettime()` |
| **8** | The loop pops and executes the `Handle`. The callback marks the `sleep`'s internal future as "done". | `deque([])` | `[]` | `PENDING` | `handle._run()` | - |
| **9** | The `Future`'s done-callback runs, adding `Handle(Task('my_main')._step)` to `_ready` to schedule the task's resumption. | `deque([Handle(Task('my_main')._step)])` | `[]` | `PENDING` | Inside `future.set_result`. | - |
| **10**| The loop pops `Handle(Task('my_main')._step)` and runs it. The `my_main` coroutine is resumed. | `deque([])` | `[]` | `RUNNING` | `print(...)` runs. | - |
| **11**| `my_main` returns. `Task('my_main')` is finished. `run_until_complete` stops the loop. | `deque([])` | `[]` | `FINISHED` (`_state='FINISHED'`) | `asyncio.run()` cleans up. | - |

---

### 5. Detailed Execution Flow: Network I/O

#### Definitions for UNIX Networking

*   **Socket:** An abstraction representing one endpoint of a two-way communication link between two programs on a network. In UNIX, a socket is created with the `socket()` syscall.
*   **File Descriptor (FD):** A non-negative integer that serves as a handle or reference for an open resource within a process, managed by the OS kernel. When you create a socket with `socket()`, the kernel returns a file descriptor for it. Subsequent operations on that socket (like `bind`, `listen`, `read`, `write`) all use this FD to tell the kernel which resource to act upon. In UNIX, "everything is a file," so sockets are treated like files and accessed via FDs.

**Code Context:**
```python
import asyncio
from asyncio import StreamReader, StreamWriter

async def handle_echo(reader: StreamReader, writer: StreamWriter) -> None:
    print("Connection accepted.")
    data: bytes = await reader.read(100)
    writer.write(data)
    await writer.drain()
    writer.close()
    await writer.wait_closed()
    print("Connection closed.")

async def main() -> None:
    server = await asyncio.start_server(handle_echo, '127.0.0.1', 8888)
    async with server:
        await server.serve_forever()

asyncio.run(main())
```

| Step | Action | `loop._ready` | `loop._scheduled` | `loop._selector` Watch List | Task State/Properties | Syscalls Used |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **1** | `asyncio.run(main())` starts `Task('main')`. It hits `await asyncio.start_server(...)`. | `deque([])` | `[]` | `{}` | `RUNNING` (`Task('main')`) | - |
| **2** | `start_server` calls `loop.add_reader(listener_fd, _accept_connection_cb)`. | `deque([])` | `[]` | `{ listener_fd: (_accept_cb, ...) }`| `RUNNING` | `socket()`, `bind()`, `listen()`, `epoll_ctl(ADD)` |
| **3** | `start_server` returns. `Task('main')` hits `await server.serve_forever()`. It **yields**, waiting on an internal future that never completes. | `deque([])` | `[]` | `{ listener_fd: ...}` | `PENDING` (`Task('main')`) | - |
| **4** | Loop is idle, **blocked** in `selector.select()`. | `deque([])` | `[]` | `{ listener_fd: ...}` | - | `epoll_wait(timeout=-1)` |
| **5** | Client connects. `epoll_wait` **returns immediately** with `listener_fd`. `_process_events` adds `Handle(_accept_cb)` to `_ready`. | `deque([Handle(_accept_cb)])` | `[]` | `{...}` | - | - |
| **6** | Loop runs `Handle(_accept_cb)`. It calls `accept()`, gets `client_fd`, and creates a `Transport` (`_SelectorSocketTransport`) and `Protocol` (`StreamReaderProtocol`). | `deque([])` | `[]` | `{...}` | - | `accept()` |
| **7** | The new `Transport` calls `loop.add_reader(client_fd, protocol.feed_data)`. `feed_data` is the callback. A new `Task('echo')` is created for the `handle_echo` coroutine and put on `_ready`. | `deque([Handle(Task('echo')._step)])` | `[]` | `{..., client_fd: (feed_data, ...)}` | `PENDING` (`Task('echo')`) | `epoll_ctl(ADD)` |
| **8** | Loop runs `Task('echo')`. It prints, then hits `data = await reader.read(100)`. `read()` sees no data is buffered. | `deque([])` | `[]` | `{...}` | `RUNNING` (`Task('echo')`) | - |
| **9** | `reader.read()` returns a new, empty `Future`. `Task('echo')` **yields**, waiting for this `Future` to be resolved by the protocol. | `deque([])` | `[]` | `{..., client_fd: ...}` | `PENDING` (`Task('echo')`) | `epoll_wait()` |
| **10**| Client sends data. `epoll_wait` returns with `client_fd`. `_process_events` finds the `feed_data` callback and adds `Handle(feed_data)` to `_ready`. |`deque([Handle(feed_data)])`|`[]`|`{...}`|`PENDING`| - |
| **11**| Loop runs `Handle(feed_data)`. The callback calls `sock.recv()` and passes the data to the `StreamReaderProtocol`, which resolves the `Future` from step 9. This triggers the callback to resume `Task('echo')`. |`deque([Handle(Task('echo')._step)])`|`[]`|`{...}`|`PENDING`| `recv()` |
| **12**| Loop runs `Handle(Task('echo')._step)`. `Task('echo')` resumes. `await reader.read()` returns the data. The task finishes its logic and returns. |`deque([])`|`[]`|`{...}`|`FINISHED` (`Task('echo')`)| `write()`, `close()` |

---

### 6. Detailed Execution Flow: File I/O (Thread Pool)

**Code Context:**
```python
import asyncio
import aiofiles

async def my_handler() -> None:
    async with aiofiles.open('file.txt', mode='r') as f:
        # The 'await' is implicit in 'async with' aiofiles
        contents: str = await f.read()

asyncio.run(my_handler())
```

| Step | Action | `loop._ready` | `loop._scheduled` | `Task('handler')` State/Properties | Threads / Syscalls |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **1** | `asyncio.run()` starts `Task('handler')`. It hits `aiofiles.open(...)`. | `deque([])` | `[]` | `RUNNING` (`_state='RUNNING'`) | Main Thread |
| **2** | `aiofiles` knows `open` is blocking, so it calls `future = loop.run_in_executor(None, open, 'file.txt', 'r')`. | `deque([])` | `[]` | `RUNNING` | `ThreadPoolExecutor` gets the `open` job. |
| **3** | `Task('handler')` **yields**, awaiting the `Future(F)` from `run_in_executor`. It attaches a callback to `Future(F)` to schedule its own resumption. | `deque([])` | `[]` | `PENDING` (`_state='PENDING'`) | A Worker Thread is dispatched from the pool. |
| **4.1**| **Main Thread** is free to run the event loop. It finds nothing to do and blocks in `selector.select()`. | `deque([])` | `[]` | `PENDING` | Main Thread `epoll_wait()` |
| **4.2**| **Worker Thread** is **blocked** inside the `open()` syscall, waiting for the disk. | - | - | - | Worker Thread calls blocking `open()` |
| **5** | The `open()` syscall completes. The worker thread has the file object result. | - | - | - | Worker Thread |
| **6** | The worker thread's executor wrapper calls `future.set_result(file_obj)` on `Future(F)`. | - | - | `PENDING` | Worker Thread |
| **7** | `set_result` triggers the attached done-callback. From the worker thread, it calls `loop.call_soon_threadsafe(Task('handler')._step)`. | `deque([Handle(Task('handler')._step)])` | `[]` | `PENDING` | `write()` to loop's self-pipe wakes main thread. |
| **8** | Main thread's loop finds and runs `Handle(Task('handler')._step)`. | `deque([])` | `[]` | `RUNNING` | Main Thread |
| **9** | `Task('handler')` resumes inside the `async with`. It then hits `await f.read()`, which triggers this entire cycle (Steps 2-8) again for the `read` operation. After the second resumption, it returns. | `deque([])` | `[]` | `FINISHED` (`_state='FINISHED'`) | Main Thread |
